### ISSUE TITLE
Parallel Linear System Solver Iterations Increases with Core Count

### ISSUE THREAD

**Issue by robert-chiodi** [mfem/mfem#2327]
Hi all,

I am developing a hypersonic vehicle material-response solver based on MFEM and using DG. As part of this system, I need to solve a parabolic PDE for the pressure that appears as:

<img width="605" alt="Screen Shot 2021-06-11 at 1 29 38 PM" src="https://user-images.githubusercontent.com/47005129/121745382-14b37d80-cab9-11eb-9b88-4df065a27f6e.png">

The equation is discretized in time using Crank-Nicolson, leading to an implicit solve. All of this seems to be working well and I can recover results from literature for the system.

I have noticed, however, that as I increase the number of MPI ranks used to perform the simulation, the number of iterations to solve the linear system rapidly increases. For example, in serial (1 rank), the linear-system is solved in 1 iteration. With 4 ranks it takes ~14 iterations to solve, and with 8 ranks it takes around ~32 iterations. The system is being solved with mfem::GMRES, and similar scaling of the iterations with the number of ranks is apparent when using BlockILU as a preconditioner or the AIR preconditioner from ex9p. The system is being solved to a relative tolerance of 1.0e-7 and in all cases the correct answer is ultimately recovered.

My main question is if this is expected behavior or if it indicates I have a bug in my code somewhere? Any advice on either pinpointing the issue or addressing it would be greatly appreciated.

Thank you

**Comment by pazner** [comment]
Hi @robert-chiodi,

* The block ILU preconditioner is uncoupled across MPI ranks, and so it is expected that as your increase the number of MPI ranks, the iteration counts would increase. That being said, the iterations you are reporting are increasing faster than I would have expected.
* AIR will probably be more scalable in parallel if you are targeting very large number of ranks, and it is not expected to see such extreme degradation as you increase the number of ranks (cc. @bensworth).
* Do you know why your solver is converging in only one iteration when you run with one rank? This is not what I would expect for a parabolic-type problem. Both AIR and block ILU will solve the problem exactly (on one rank) if it can be reordered to obtain a (block) triangular system (for example, when you are solving a purely advective problem with no cycles, and so a sweeping algorithm is possible). Block ILU is exact in some other limited cases (when there is no fill-in, and the greedy minimum discarded fill can identify this), but for standard parabolic problems this is not the case.

**Comment by bensworth** [comment]
Agree with Will on the one iteration question. Also, when you increase processors is this a weak or strong style scaling? I.e., are you increasing problem size or just processor count? And how big is the problem in terms of DOFs? As long as the DOFs aren't such that each processor has a very small number, AIR should be fairly robust in terms of fixing problem and increasing problem size, with the one exception of a fully hyperbolic and upwinded problem, related to Will's last point.

Also, how are you linearizing, and do you have any notion of the ratio of advection to diffusion in the linearized system? Unless the advection is fairly strong, normal AMG may be a good bet.

**Comment by robert-chiodi** [comment]
Thank you both for your quick responses. I will try my best to answer the questions (and apologies if I end up being overly thorough).

I am unsure of why it is solving the linear system in only one iteration when simulated in serial. The example results I gave were under strong-scaling with 400 DOFs making up the pressure field. However, I have experienced the same behavior under strong scaling with 400,000 DOFs and weak-scaling with 40,000 DOFs per core. I tried HypreBoomerAMG before and a similar behavior was observed. I should mention that I am solving a 1D problem (for now) in a 2D domain. I am unsure if that is relevant for this question.

The discretized equation I am solving is 
<img width="456" alt="Screen Shot 2021-06-11 at 5 20 00 PM" src="https://user-images.githubusercontent.com/47005129/121759297-e1cdb180-cad9-11eb-952b-0f760c342490.png">

with each term being
<img width="325" alt="Screen Shot 2021-06-11 at 5 24 55 PM" src="https://user-images.githubusercontent.com/47005129/121759315-f3af5480-cad9-11eb-938d-ea833eeb422a.png">
<img width="470" alt="Screen Shot 2021-06-11 at 5 25 13 PM" src="https://user-images.githubusercontent.com/47005129/121759337-12155000-cada-11eb-80c4-4e6954e6d64a.png">
<img width="343" alt="Screen Shot 2021-06-11 at 5 25 16 PM" src="https://user-images.githubusercontent.com/47005129/121759340-15a8d700-cada-11eb-9486-fefd515364dd.png">
<img width="451" alt="Screen Shot 2021-06-11 at 5 25 20 PM" src="https://user-images.githubusercontent.com/47005129/121759343-1b9eb800-cada-11eb-8cf6-64bca9a4178e.png">
<img width="464" alt="Screen Shot 2021-06-11 at 5 25 24 PM" src="https://user-images.githubusercontent.com/47005129/121759344-1e99a880-cada-11eb-8db5-1991e8d6b39f.png">
<img width="251" alt="Screen Shot 2021-06-11 at 5 25 28 PM" src="https://user-images.githubusercontent.com/47005129/121759346-21949900-cada-11eb-9b8e-05d1b6f1ec20.png">
<img width="248" alt="Screen Shot 2021-06-11 at 5 25 31 PM" src="https://user-images.githubusercontent.com/47005129/121759352-25282000-cada-11eb-80b0-9167e7a2cd06.png">

The DGDiffusionIntegrator is used for `D_I` term to lead to SIPG, with sigma = -1.0 and kappa = (order+1)^2, with the order being p = 1 in this case.

For a relative order of magnitude of the coefficients involved here, the term inside the `M` term has an average value of 6.43553e-06 and min/max values of 7.11182e-07/7.05651e-06 over the domain. Similarly, the coefficient term for `D_I` has an average value of 2.53237e-09 and min/max values of 8.55904e-11/2.86114e-09.

Lastly, here is the convergence of the preconditioned GMRES solver (BlockILU preconditioner with `MINIMUM_DISCARDED_FILL`) at the 1000th iteration of my simulation (for which the above coefficient values are also given.)

```
1 Core, at 1000th iteration
   Pass :  1   Iteration :   0  ||B r|| = 1817.42
   Pass :  1   Iteration :   1  ||B r|| = 8.82305e-12
```

```
4 Cores, at 1000th iteration
   Pass :  1   Iteration :   0  ||B r|| = 1738.76
   Pass :  1   Iteration :   1  ||B r|| = 1.20532
   Pass :  1   Iteration :   2  ||B r|| = 1.11088
   Pass :  1   Iteration :   3  ||B r|| = 0.661451
   Pass :  1   Iteration :   4  ||B r|| = 0.372865
   Pass :  1   Iteration :   5  ||B r|| = 0.369778
   Pass :  1   Iteration :   6  ||B r|| = 0.368401
   Pass :  1   Iteration :   7  ||B r|| = 0.363899
   Pass :  1   Iteration :   8  ||B r|| = 0.360779
   Pass :  1   Iteration :   9  ||B r|| = 0.300359
   Pass :  1   Iteration :  10  ||B r|| = 0.290819
   Pass :  1   Iteration :  11  ||B r|| = 0.277159
   Pass :  1   Iteration :  12  ||B r|| = 0.262182
   Pass :  1   Iteration :  13  ||B r|| = 0.0410115
   Pass :  1   Iteration :  14  ||B r|| = 0.0382026
   Pass :  1   Iteration :  15  ||B r|| = 0.0378235
   Pass :  1   Iteration :  16  ||B r|| = 0.00617479
   Pass :  1   Iteration :  17  ||B r|| = 5.42205e-05
```

```
8 Cores, at 1000th iteration
   Pass :  1   Iteration :   0  ||B r|| = 1498.19
   Pass :  1   Iteration :   1  ||B r|| = 470.079
   Pass :  1   Iteration :   2  ||B r|| = 64.7397
   Pass :  1   Iteration :   3  ||B r|| = 63.1177
   Pass :  1   Iteration :   4  ||B r|| = 13.4724
   Pass :  1   Iteration :   5  ||B r|| = 2.44903
   Pass :  1   Iteration :   6  ||B r|| = 0.827178
   Pass :  1   Iteration :   7  ||B r|| = 0.827089
   Pass :  1   Iteration :   8  ||B r|| = 0.527957
   Pass :  1   Iteration :   9  ||B r|| = 0.524465
   Pass :  1   Iteration :  10  ||B r|| = 0.524447
   Pass :  1   Iteration :  11  ||B r|| = 0.399499
   Pass :  1   Iteration :  12  ||B r|| = 0.394473
   Pass :  1   Iteration :  13  ||B r|| = 0.393724
   Pass :  1   Iteration :  14  ||B r|| = 0.30696
   Pass :  1   Iteration :  15  ||B r|| = 0.306346
   Pass :  1   Iteration :  16  ||B r|| = 0.304709
   Pass :  1   Iteration :  17  ||B r|| = 0.246696
   Pass :  1   Iteration :  18  ||B r|| = 0.243224
   Pass :  1   Iteration :  19  ||B r|| = 0.243033
   Pass :  1   Iteration :  20  ||B r|| = 0.218784
   Pass :  1   Iteration :  21  ||B r|| = 0.215724
   Pass :  1   Iteration :  22  ||B r|| = 0.214978
   Pass :  1   Iteration :  23  ||B r|| = 0.198691
   Pass :  1   Iteration :  24  ||B r|| = 0.191974
   Pass :  1   Iteration :  25  ||B r|| = 0.185555
   Pass :  1   Iteration :  26  ||B r|| = 0.155799
   Pass :  1   Iteration :  27  ||B r|| = 0.148799
   Pass :  1   Iteration :  28  ||B r|| = 0.117595
   Pass :  1   Iteration :  29  ||B r|| = 0.0990408
   Pass :  1   Iteration :  30  ||B r|| = 0.0958863
   Pass :  1   Iteration :  31  ||B r|| = 0.00594816
   Pass :  1   Iteration :  32  ||B r|| = 0.00210528
   Pass :  1   Iteration :  33  ||B r|| = 0.00194718
   Pass :  1   Iteration :  34  ||B r|| = 0.000721814
   Pass :  1   Iteration :  35  ||B r|| = 0.000243363
   Pass :  1   Iteration :  36  ||B r|| = 1.73888e-06
```

I hope some of this information ends up being helpful. Please feel free to ask more questions too, including the same ones in case I did not answer them.

**Comment by pazner** [comment]
> I should mention that I am solving a 1D problem (for now) in a 2D domain. I am unsure if that is relevant for this question.

This actually explains why you converge in one iteration in serial: in 1D, your matrix will be (block) tridiagonal, and so there will be no fill-in and the solve will be exact.

It could be that your system is quite ill-conditioned and that it is not very well preconditioned by ILU/AIR/AMG, and the only reason you see good convergence on a small number of processors is because the solver is close to being direct. As Ben mentioned, if diffusion dominates, standard AMG might be a better bet.

Sometimes DG requires specialized preconditioners (e.g. in situations when the IP penalty parameter is large). That might be worth looking into in your case, but it will likely be problem-specific.

**Comment by robert-chiodi** [comment]
That makes a lot of sense, thank you so much for your help. I believe you were 100% correct. I made the case fully two-dimensional and the convergence behavior is making a lot more sense now. Indeed, standard AMG does seem to be a better preconditioner in this case since it is only diffusion. 

I'll mark the issue closed, and again, thank you very much @pazner and @bensworth for your help.

**Comment by bensworth** [comment]
@robert-chiodi Also worth pointing out: posing a 1d problem on a 2d mesh can actually cause serious problems for solvers. You have effectively posed a totally anisotropic diffusion problem, with singular diffusion coefficient in 2d. Just something to be aware of in case you consider such a workaround again.

**Comment by robert-chiodi** [comment]
@bensworth Thanks indeed for pointing that out. I was aware it was a degenerate case, however, I did not know how serious of a problem it could create. I have updated my code to allow running fully in 1D. It is really neat how easy it was to do due to MFEM.

As a side note, I think 1D integration rules do not currently set their `IntegrationPoint` `index` member variables for the `IntegrationPoint` objects held in `IntegrationRule`. I currently use some `QuadratureFunctionCoefficients`, and saw that, when switching to 1D, calls to `QuadratureFunction::GetElementValues` could lead to segfaults. After tracking down the issue, the index value seemed to be unset, which could cause an erroneous `s_offset` value leading to the segfault. I fixed it by calling `ir->SetPointIndices()` at the bottom of `IntegrationRules::SegmentIntegrationRule` before returning. Should I open a new GitHub issue for this?

**Comment by pazner** [comment]
> As a side note, I think 1D integration rules do not currently set their `IntegrationPoint` `index` member variables for the `IntegrationPoint` objects held in `IntegrationRule`. I currently use some `QuadratureFunctionCoefficients`, and saw that, when switching to 1D, calls to `QuadratureFunction::GetElementValues` could lead to segfaults. After tracking down the issue, the index value seemed to be unset, which could cause an erroneous `s_offset` value leading to the segfault. I fixed it by calling `ir->SetPointIndices()` at the bottom of `IntegrationRules::SegmentIntegrationRule` before returning. Should I open a new GitHub issue for this?

Yes, please open a new issue for this. Thanks!
